add graylog compatibility

negbie opened the pull request.

negbie committed:
append graylog connection

negbie committed:
Update kamailio.cfg

negbie committed:
Update kamailio.cfg

negbie committed:
Update kamailio.cfg

lmangani commented on data/kamailio.cfg:
```diff
@@ -1268,9 +1552,13 @@ route[SEND_ES_STATS] {
 
 	# Index+Type defined in bulk object
-	$var(elasticurl) = "ELASTICSEARCH_URL/_bulk";
+	$var(elasticurl) = "http://YOUR_GRAYLOG_IP:12201/gelf";
```
should $var(elasticurl) also be conditional and retain support both GRAYLOG and existing ES?

negbie replied:
definitely

negbie committed:
conditional graylog url

negbie committed:
fix typo

grumvalski commented on data/kamailio.cfg:
```diff
@@ -1267,10 +1552,14 @@ route[SEND_ES_STATS] {
 	# Index+Type defined in bulk object
-	$var(elasticurl) = "ELASTICSEARCH_URL/_bulk";
-
+	$var(elasticurl) = "http://YOUR_ELASTIC_IP:9200/_bulk";
```
In this way you are hardcoding the protocol and the port on which elasticsearch (or graylog) run. I still think here is better to have the full url which could be composed elsewhere by configuration variables (e.g.: transport, ip, port).

negbie replied:
Good point. Will change that!

negbie committed:
Global GRAYLOG_GELF_HTTP_URL

negbie committed:
Update comment

negbie committed:
fix typo

negbie committed:
support for sending alarms to graylog

negbie committed:
add useragent field for scanner

negbie committed:
changed seperation string in scanner hashtable

Commit:
resolve merge conflict

negbie committed:
Add X-RTP-Stat parsing

negbie committed:
hashtable for x-rtp-stats

negbie committed:
clean up

negbie committed:
fix sending methods in check_stats route

negbie committed:
Add mos values*10^3

To circumvent floats, the mos values are shifted by 10^3

negbie committed:
Ups mixed up local file

negbie committed:
replaced s.select with re.subst

negbie committed:
i hate spaces ;)

negbie committed:
still hating spaces!

grumvalski commented on data/kamailio.cfg:
```diff
@@ -586,17 +770,29 @@ route[CHECK_ALARM]
+						$var(full) = $sel(msg);
+						jansson_set("string", "short_message", ""+$var(short)+"", "$var(tmp_json)");
+						jansson_set("string", "full_message", ""+$var(full)+"", "$var(tmp_json)");
+						$var(graylog_json) = $var(tmp_json);
```
This part looks to me repeated in the script. Wouldn't it be better to put it in the SEND_ES_DATA route inside the 
#!ifdef TO_GRAYLOG?
So, using jansson_set on the $var(body).

negbie replied:
I am still not confident with my field naming. the full_message field is optional and maybe i will remove it completly. short_message is mandatory for gelf. if i decide to keep it static i will move it to SEND_ES_DATA.

grumvalski commented on data/kamailio.cfg:
```diff
@@ -768,8 +984,20 @@ route[CHECK_STATS] {
 		$sht(a=>packet::count) = 0;
+
+		$var(body) = $var(stats_data);
+		route(SEND_ES_DATA);
```
Why did you split data sending in this way. In this way you are basically sending an HTTP request per document (elasticsearch or graylog), loosing the benefit of the _bulk API. For this reason the stats where grouped by type.

negbie replied:
For graylog you cannot send directly to es. You need to spawn and gelf http input and i dont know if this input type supports the _bulk api. I need to google it first or ask in the user group.

negbie replied:
But nevertheless i will change it to get a similar behaviour to elastic

grumvalski commented on data/kamailio.cfg:
```diff
@@ -353,129 +364,289 @@ onreply_route {
+		}
+	}
+}
+
```
Really nice one. I'd like to have the x-rtp stats on elasticsearch too. Could we generalize this part in order to follow the same logic as per the other stats (so valid both for elasticsearch and graylog).

negbie replied:
will add it.

grumvalski reviewed (commented):
Finally I could have a look on this PR. Overall good, some remarks.

negbie committed:
add x-rtp-stats for elasticsearch

grumvalski commented on data/kamailio.cfg:
```diff
@@ -537,113 +538,121 @@ route[PARSE_X_RTP_STATS] {
+		$sht(x=>$ci::callid_count) = $sht(x=>$ci::callid_count) + 1;
+		
+		$var(body) = $var(xrtp_stats);
+		route(SEND_ES_DATA);
```
thanks :)

negbie replied:
Your welcome i still need to add but will do in another pr

jansson_set("string", "host", "$var(host)", "$var(tmp_json)");
jansson_set("string", "timestamp", "$var(isotime)", "$var(tmp_json)");

lmangani commented:
Awesome job guys! Are we looking good for a merge or shall we modularize this further?

negbie commented:
Lets wait a bit till i get a final answer from here:
https://groups.google.com/forum/#!topic/graylog2/nTr8vFdlizc
Still then i need to polish a bit ;)

grumvalski commented:
two points still to address imho:
1) code duplication for graylog logs
2) bulk/aggregated requests

negbie commented:
1) true but maybe i will let it inside the #!ifdef TO_GRAYLOG. I mean it does not hurt there anybody and i will need that ifdef anyway because i need another timestamp for graylog "timestamp", ""+$TS+"."+$TV(u)/999+""
One advantage is when you need additional information for example you want the whole message body for 50X response codes you can easily add it to short_message or full_message without add extra fields. In graylog you can easily extract those information.

2) it seems graylog cannot handle it (yet) in the same way like sending it directly to elastic. So i will need route(SEND_ES_DATA) for every request but i will move it to the ifdef so it will affect only graylog users.

grumvalski commented:
Understand. 
We'll think about refactoring and ave more compact code later :)

negbie commented:
Btw. i noticed that we we need to take care about the private memory pool space. run sometimes into "not enough memory to initialize destination branches" errors.

Maybe we should raise PKG_MEMORY in /etc/default/kamailio to 16 or make ifdefs for elasticsearch and graylog.

negbie committed:
keep aggregated requests for elastic

negbie committed:
opt out graylog

grumvalski commented:
This is strange because I don't think that we are using so much the package memory, most of the structure being in shm. Do you see which child is printing the rror: maybe the http async worker(s)? When this occur is it something that stay "stable" or the errors are only from time to time? could you run
kamcmd pkg.stats
kamcmd mod.stats all pkg

when this happens?

negbie commented:
Ok it seems there is no memory leak but just not enough memory during the startup of kamailio. When i remove some private memory variables till kamailio is able to start you have more or less those stats for all the workers:

{
        entry: 10
        pid: 19936
        rank: -2
        used: 4542944
        free: 2951136
        real_used: 5437472
        total_size: 8388608
        total_frags: 1464
}

negbie commented:
It comes from the regular expressions which i use a lot in route[PARSE_X_RTP_STATS]

negbie committed:
macro for elasticsearch and graylog

Added a macro for send data to elastic search or graylog.
Added some extra methods and 488 response code.

Private memory usage per worker with:
#!define WITH_STATISTIC_METHOD_EXTRA
#!define WITH_STATISTIC_INVITE_1XX
#!define TO_GRAYLOG

Around 3.1 MB

Private memory usage per worker with:
#!define WITH_STATISTIC_METHOD_EXTRA
#!define WITH_STATISTIC_INVITE_1XX
#!define TO_ELASTIC

Around 3.4 MB

#!define WITH_X_RTP_STATS will need around 0.61 MB

grumvalski commented on data/kamailio.cfg:
```diff
@@ -1306,87 +2086,100 @@ route[CHECK_STATS] {
 
-	# Index+Type defined in bulk object
+#!ifdef TO_GRAYLOG
+	if ($var(graylog_json) == $null || $var(graylog_json) == 0 ) return;
```
a $var cannot be $null in kamailio, either is "" or 0.

grumvalski commented on data/kamailio.cfg:
```diff
@@ -1306,87 +2086,100 @@ route[CHECK_STATS] {
 #!endif
 }
 
 route[SEND_ES_DATA] {
```
maybe we could rename this route to something like SEND_HTTP_DATA (and the reply accordingly).

negbie replied:
makes sense.

grumvalski commented:
Looks good. I think at this point we should merge, what do you think @lmangani ?

negbie committed:
$var cannot be $null, either is "" or 0.

grumvalski commented:
let's merge it!
